sstables/trie/trie_writer.hh

```diff
@@ -0,0 +1,1192 @@
+    { co.pos() } -> std::same_as<sink_pos>;
+    { co.page_size() } -> std::same_as<uint64_t>;
+    { co.bytes_left_in_page() } -> std::same_as<uint64_t>;
+};
+
+// The trie writer holds a trie of these nodes while it's building the index pages.
+//
+// Whenever a new key is fed to the writer:
+// 1. The "latest node" pointer climbs up the rightmost path to the point of mismatch between
+// the previous key and the new key, and "completes" nodes along the way.
+// (A node is "completed" when we climb above it, which means that it won't receive any new children).
+// 2. If a newly-completed subtree can't fit into a page of the file writer,
+// branches of the subtree are flushed and freed from memory,
+// and only the root of the subtree remains in memory, with metadata updated accordingly.
+// 3. A new node, containing the chain of new characters after the mismatch point, is added as the latest node.
+struct writer_node {
+    // The payload of this node. It might be absent -- this is represented by _payload._payload_bits == 0.
+    //
+    // A node has a payload iff it corresponds to some key.
+    // All leaf nodes must contain a payload.
+    trie_payload _payload;
+    // True iff any of the direct children of this node lies on a different page
+    // (i.e. has been written out beforehand, in a different lay_out_children() than this node).
+    // It means that the size of this node might change from the initial estimate at completion
+    // (because the offset to children might have grow, and the offset integers need more bits to fit),
+    // so it has to be recalculated before the actual write.
+    //
+    // After the node itself is written out, this is reset to `false`.
+    //
+    // Invalid after _pos is set.
+    bool _has_out_of_page_children = false;
+    // True iff some yet-unwritten descendant has out-of-page children.
+    // This means that the sizes of our children (and thus also our size) might change from the initial
+    // estimate at completion time (because some offsets in their subtree might have grown, and the offset
+    // integers now require more bits), so they have to be recalculated before the actual write.
+    //
+    // After the node's children are written out, this is always `false`.
+    //
+    // Invalid after _pos is set.
+    bool _has_out_of_page_descendants = false;
+    // This is a byte-based trie. A path from the root to a node corresponds
+    // to some key, with each successive node appending several characters.
+    // _transition points at the chain of characters (of length _transition_meta)
+    // corresponding to this node.
+    //
+    // In the root node, _transition is meaningless, though it's still set to something non-empty.
+    //
+    // In uncompleted nodes, must be allocated after the node itself.
+    // (Our memory management assumes a certain order of things inside the bump_allocator).
+    //
+    // Invalid after _pos is set.
+    ptr<std::byte[]> _transition;
```

Comment:
Would it make sense to allocate it locally here (and limit it to some small amount?)